File: src/main/java/org/snipsniper/snipscope/SnipScopeListener.java
package org.snipsniper.snipscope;

import org.snipsniper.snipscope.ui.SnipScopeUIComponent;
import org.snipsniper.utils.InputContainer;
import org.snipsniper.utils.Vector2Int;

import java.awt.*;
import java.awt.event.*;

public class SnipScopeListener implements KeyListener, MouseListener, MouseMotionListener, MouseWheelListener{
    private final SnipScopeWindow snipScopeWindow;
    private final InputContainer input;
    private Point lastPoint;

    public SnipScopeListener(SnipScopeWindow snipScopeWindow) {
        this.snipScopeWindow = snipScopeWindow;
        input = snipScopeWindow.getInputContainer();
        //TODO: Why is this here? Can we move this?
        snipScopeWindow.addComponentListener(new ComponentAdapter() {
            public void componentResized(ComponentEvent evt) {
                snipScopeWindow.resizeTrigger();
            }
        });
    }

    //Key listener

    @Override
    public void keyTyped(KeyEvent keyEvent) { }

    @Override
    public void keyPressed(KeyEvent keyEvent) {
        input.setKey(keyEvent.getKeyCode(), true);

        if(!snipScopeWindow.isEnableInteraction()) return;

        switch(keyEvent.getKeyCode()) {
            case KeyEvent.VK_R: snipScopeWindow.resetZoom(); break;
            case KeyEvent.VK_ESCAPE: snipScopeWindow.dispose(); break;
        }

        snipScopeWindow.repaint();
    }

    @Override
    public void keyReleased(KeyEvent keyEvent) {
        input.setKey(keyEvent.getKeyCode(), false);
    }

    //Mouse listener

    @Override
    public void mouseClicked(MouseEvent mouseEvent) { }

    @Override
    public void mousePressed(MouseEvent mouseEvent) {
        if(!snipScopeWindow.isEnableInteraction()) return;

        for(SnipScopeUIComponent component : snipScopeWindow.getUiComponents()) {
            component.mousePressed(mouseEvent);
        }
    }

    @Override
    public void mouseReleased(MouseEvent mouseEvent) {
        if(!snipScopeWindow.isEnableInteraction()) return;

        lastPoint = null;
        for(SnipScopeUIComponent component : snipScopeWindow.getUiComponents()) {
            component.mouseReleased(mouseEvent);
        }
    }

    @Override
    public void mouseEntered(MouseEvent mouseEvent) { }

    @Override
    public void mouseExited(MouseEvent mouseEvent) { }

    //Mouse motion listener

    @Override
    public void mouseDragged(MouseEvent mouseEvent) {
        if(!snipScopeWindow.isEnableInteraction()) return;

        for(SnipScopeUIComponent component : snipScopeWindow.getUiComponents()) {
            component.mouseDragged(mouseEvent);
        }

        if (input.isKeyPressed(snipScopeWindow.getMovementKey())) {
            if(lastPoint == null) lastPoint = mouseEvent.getPoint();
            double x = lastPoint.getX() - mouseEvent.getPoint().getX();
            double y = lastPoint.getY() - mouseEvent.getPoint().getY();
            snipScopeWindow.setPosition(Vector2Int.add(snipScopeWindow.getPosition(), new Vector2Int(x, y)));

            lastPoint = mouseEvent.getPoint();
            snipScopeWindow.repaint();
        }
    }

    @Override
    public void mouseMoved(MouseEvent mouseEvent) {
        if(!snipScopeWindow.isEnableInteraction()) return;

        for(SnipScopeUIComponent component : snipScopeWindow.getUiComponents()) {
            component.mouseMoved(mouseEvent);
        }
    }

    //Mouse wheel listener

    @Override
    public void mouseWheelMoved(MouseWheelEvent mouseWheelEvent) {
        if(!snipScopeWindow.isEnableInteraction()) return;

        if(!snipScopeWindow.isRequireMovementKeyForZoom() || input.isKeyPressed(snipScopeWindow.getMovementKey())) {
            float oldZoom = snipScopeWindow.getZoom();
            switch (mouseWheelEvent.getWheelRotation()) {
                case -1: snipScopeWindow.setZoom(oldZoom + 0.1F); break;
                case 1: if (oldZoom >= 0.2F) snipScopeWindow.setZoom(oldZoom - 0.1F); break;
            }
            snipScopeWindow.calculateZoom();
        }
    }
}

File: src/main/kotlin/net/snipsniper/snipscope/SnipScopeListener.kt
package net.snipsniper.snipscope

import net.snipsniper.utils.InputContainer
import org.capturecoop.ccutils.math.CCVector2Int
import java.awt.Point
import java.awt.event.*

open class SnipScopeListener(private val snipScopeWindow: SnipScopeWindow): KeyListener, MouseListener, MouseMotionListener, MouseWheelListener {
    private val input: InputContainer = snipScopeWindow.inputContainer
    private var lastPoint: Point? = null

    init {
        //TODO: Why is this here? Can we move this?
        snipScopeWindow.addComponentListener(object: ComponentAdapter() {
            override fun componentResized(e: ComponentEvent) {
                snipScopeWindow.resizeTrigger()
            }
        })
    }

    //Key Listener

    override fun keyTyped(keyEvent: KeyEvent) { }

    override fun keyPressed(keyEvent: KeyEvent) {
        input.setKey(keyEvent.keyCode, true)
        if(!snipScopeWindow.isEnableInteraction) return
        when(keyEvent.keyCode) {
            KeyEvent.VK_R -> snipScopeWindow.resetZoom()
            KeyEvent.VK_ESCAPE -> snipScopeWindow.dispose() //TODO: Make this optional or something
        }
        snipScopeWindow.repaint()
    }

    override fun keyReleased(keyEvent: KeyEvent) {
        input.setKey(keyEvent.keyCode, false)
    }

    //Mouse Listener

    override fun mouseClicked(mouseEvent: MouseEvent) { }

    override fun mousePressed(mouseEvent: MouseEvent) {
        if(!snipScopeWindow.isEnableInteraction) return
        snipScopeWindow.uiComponents.forEach { it.mousePressed(mouseEvent) }
    }

    override fun mouseReleased(mouseEvent: MouseEvent) {
        if(!snipScopeWindow.isEnableInteraction) return
        lastPoint = null
        snipScopeWindow.uiComponents.forEach { it.mouseReleased(mouseEvent) }
    }

    override fun mouseEntered(mouseEvent: MouseEvent) { }

    override fun mouseExited(mouseEvent: MouseEvent) { }

    //Mouse Motion Listener

    override fun mouseDragged(mouseEvent: MouseEvent) {
        if(!snipScopeWindow.isEnableInteraction) return
        snipScopeWindow.uiComponents.forEach { it.mouseDragged(mouseEvent) }

        if (input.isKeyPressed(snipScopeWindow.movementKey)) {
            if(lastPoint == null) lastPoint = mouseEvent.point
            lastPoint?.let {
                val x = lastPoint!!.getX() - mouseEvent.point.getX()
                val y = lastPoint!!.getY() - mouseEvent.point.getY()
                snipScopeWindow.position += CCVector2Int(x, y)
            }
            lastPoint = mouseEvent.point
            snipScopeWindow.repaint()
        }
    }

    override fun mouseMoved(mouseEvent: MouseEvent) {
        if(!snipScopeWindow.isEnableInteraction) return
        snipScopeWindow.uiComponents.forEach { it.mouseMoved(mouseEvent) }
    }

    //Mousewheel Listener

    override fun mouseWheelMoved(mouseWheelEvent: MouseWheelEvent) {
        if(!snipScopeWindow.isEnableInteraction) return

        if(!snipScopeWindow.isRequireMovementKeyForZoom || input.isKeyPressed(snipScopeWindow.movementKey)) {
            val oldZoom = snipScopeWindow.zoom
            when (mouseWheelEvent.wheelRotation) {
                -1 -> snipScopeWindow.zoom = oldZoom + 0.1F
                1 -> if (oldZoom >= 0.2F) snipScopeWindow.zoom = oldZoom - 0.1F
            }
            snipScopeWindow.calculateZoom()
        }
    }
}